### PR TITLE
Protect UnixSocketServer::m_socket under atomic

### LINKED ISSUE #4554: data race observed when Ctrl-C eth/eth
I saw this when I sent `SIGTERM` to `eth/eth` under ThreadSanitizer.

```
⧎ ◌  18:31:43.547|p2p|b188467a…|Geth/v1.7.0-stable-6c6c7b2a/windows-amd64/go1.9  BlocksHeaders ( 1 entries)
⧎ ◌  18:31:43.548|p2p|b188467a…|Geth/v1.7.0-stable-6c6c7b2a/windows-amd64/go1.9  Ignoring unexpected blocks
⧎ ◌  18:31:43.714|p2p|c3db76ee…|Parity/v1.8.0-unstable-246b528-20170908/x86_64-linux-gnu/rustc1.20.0  BlocksHeaders ( 512 entries)
⧎ ◌  18:31:43.714|p2p|c3db76ee…|Parity/v1.8.0-unstable-246b528-20170908/x86_64-linux-gnu/rustc1.20.0  Ignoring unexpected blocks
⧎ ◌  18:31:43.862|p2p|621f2e5b…|Parity/v1.7.0-beta-5f2cabd-20170727/x86_64-linux-gnu/rustc1.18.0  BlocksHeaders ( 512 entries)
⧎ ◌  18:31:43.862|p2p|621f2e5b…|Parity/v1.7.0-beta-5f2cabd-20170727/x86_64-linux-gnu/rustc1.18.0  Ignoring unexpected blocks
==================
WARNING: ThreadSanitizer: data race (pid=2875)
  Write of size 4 at 0x7b4400570690 by main thread:
    #0 dev::UnixDomainSocketServer::StopListening() /home/yh/src/cpp-ethereum/libweb3jsonrpc/UnixSocketServer.cpp:96 (eth+0x7e125c)
    #1 ModularServer<>::StopListening() /home/yh/src/cpp-ethereum/libdevcore/../libweb3jsonrpc/ModularServer.h:94 (discriminator 1) (eth+0x56a0b3)
    #2 main /home/yh/src/cpp-ethereum/eth/main.cpp:1257 (eth+0x506722)

  Previous read of size 4 at 0x7b4400570690 by thread T8:
    #0 dev::UnixDomainSocketServer::Listen() /home/yh/src/cpp-ethereum/libweb3jsonrpc/UnixSocketServer.cpp:110 (eth+0x7e1337)
    #1 operator() /home/yh/src/cpp-ethereum/libweb3jsonrpc/IpcServerBase.cpp:51 (eth+0x7ef7cd)
    #2 void std::_Bind_simple<dev::IpcServerBase<int>::StartListening()::{lambda()#1} ()>::_M_invoke<>(std::_Index_tuple<>) /usr/bin/../lib/gcc/x86_64-linux-gnu/6.3.0/../../../../include/c++/6.3.0/functional:1390 (discriminator 2) (eth+0x7ef749)
    #3 std::_Bind_simple<dev::IpcServerBase<int>::StartListening()::{lambda()#1} ()>::operator()() /usr/bin/../lib/gcc/x86_64-linux-gnu/6.3.0/../../../../include/c++/6.3.0/functional:1380 (eth+0x7ef6f9)
    #4 std::thread::_State_impl<std::_Bind_simple<dev::IpcServerBase<int>::StartListening()::{lambda()#1} ()> >::_M_run() /usr/bin/../lib/gcc/x86_64-linux-gnu/6.3.0/../../../../include/c++/6.3.0/thread:197 (eth+0x7ef5bd)
    #5 std::error_code::default_error_condition() const ??:? (libstdc++.so.6+0xbb83e)

  As if synchronized via sleep:
    #0 nanosleep ??:? (eth+0x47cc17)
    #1 void std::this_thread::sleep_for<long, std::ratio<1l, 1000l> >(std::chrono::duration<long, std::ratio<1l, 1000l> > const&) /usr/bin/../lib/gcc/x86_64-linux-gnu/6.3.0/../../../../include/c++/6.3.0/thread:323 (discriminator 1) (eth+0x50cb0a)
    #2 stopSealingAfterXBlocks(dev::eth::Client*, unsigned int, unsigned int&) /home/yh/src/cpp-ethereum/eth/main.cpp:275 (discriminator 1) (eth+0x500ca4)
    #3 main /home/yh/src/cpp-ethereum/eth/main.cpp:1248 (eth+0x506660)

  Location is heap block of size 280 at 0x7b4400570580 allocated by main thread:
    #0 operator new(unsigned long) ??:? (eth+0x4ef983)
    #1 main /home/yh/src/cpp-ethereum/eth/main.cpp:1210 (eth+0x505fc4)

  Thread T8 (tid=2893, running) created by main thread at:
    #0 pthread_create ??:? (eth+0x463f56)
    #1 std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) ??:? (libstdc++.so.6+0xbbb54)
    #2 dev::IpcServerBase<int>::StartListening() /home/yh/src/cpp-ethereum/libweb3jsonrpc/IpcServerBase.cpp:51 (eth+0x7ed459)
    #3 dev::UnixDomainSocketServer::StartListening() /home/yh/src/cpp-ethereum/libweb3jsonrpc/UnixSocketServer.cpp:89 (eth+0x7e10f1)
    #4 main /home/yh/src/cpp-ethereum/eth/main.cpp:1212 (eth+0x506039)

SUMMARY: ThreadSanitizer: data race /home/yh/src/cpp-ethereum/libweb3jsonrpc/UnixSocketServer.cpp:96 in dev::UnixDomainSocketServer::StopListening()
==================
```

### PR DESCRIPTION
**Pull request by pirapira**
Fixes #4554.

### CODE REVIEW

File: libweb3jsonrpc/UnixSocketServer.h [hunk 1]
```diff
@@ -45,7 +46,7 @@ class UnixDomainSocketServer: public IpcServerBase<int>
 	size_t Read(int _connection, void* _data, size_t _size) override;
 
 	sockaddr_un m_address;
-	int m_socket = 0;
+	std::atomic<int> m_socket{0};
```
Comment (chfast) [inline libweb3jsonrpc/UnixSocketServer.h]: I think atomics are always initialized to default value (read 0). So explicit 0 is not needed.

File: libweb3jsonrpc/UnixSocketServer.h [hunk 2]
```diff
@@ -27,6 +27,7 @@ along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
 #include <thread>
 #include <sys/un.h>
 #include "IpcServerBase.h"
+#include <atomic>
```
Comment (chfast) [inline libweb3jsonrpc/UnixSocketServer.h]: Please also reorder includes.

File: libweb3jsonrpc/UnixSocketServer.h (hunk 1 above)
Comment (gumb0) [reply libweb3jsonrpc/UnixSocketServer.h]: They are not http://en.cppreference.com/w/cpp/atomic/atomic/atomic

**Comment by pirapira** [comment]
Reordered includes.

File: libweb3jsonrpc/UnixSocketServer.h [hunk 3]
```diff
@@ -23,10 +23,11 @@ along with cpp-ethereum.  If not, see <http://www.gnu.org/licenses/>.
 
 #if !defined(_WIN32)
 
-#include <string>
+#include "IpcServerBase.h"
+#include <atomic>
 #include <thread>
+#include <string>
```
Comment (chfast) [inline libweb3jsonrpc/UnixSocketServer.h]: I believe "t" is after "s" in most latin-based alphabets, but let it be.

Comment (pirapira) [reply libweb3jsonrpc/UnixSocketServer.h]: I didn't know you follow that ordering.

Comment (chfast) [reply libweb3jsonrpc/UnixSocketServer.h]: It looks more ordered to me.

Comment (pirapira) [reply libweb3jsonrpc/UnixSocketServer.h]: Yeah, I thought `<string>` doesn't probably include `<thread>`, but `<thread>` might include `<string>`.

**Pull request merged**